## **Income Evaluation**

### **Problem Statment**
Build a model using SVM algorithm to classify the income of an individual when given inputs about his personal lie

In [2]:
import pandas as pd
import numpy as np
import sklearn

In [3]:
#Importing the dataset
income_eval = pd.read_csv("C:/Users/levie/OneDrive/Desktop/sankyana/Datasets/data/income_evaluation.csv")
income_eval.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [4]:
#Checking for null values
income_eval.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [5]:
# Summary
income_eval.describe().T

count         mean            std      min       25%  \
age             10000.0      38.4520      13.597749     17.0      28.0   
fnlwgt          10000.0  190679.0964  106284.505611  19302.0  118492.0   
education-num   10000.0      10.0766       2.548172      1.0       9.0   
capital-gain    10000.0    1058.0274    7272.178369      0.0       0.0   
capital-loss    10000.0      87.6714     400.371885      0.0       0.0   
hours-per-week  10000.0      40.5303      12.277197      1.0      40.0   

                     50%        75%        max  
age                 37.0      47.00       90.0  
fnlwgt          179126.0  239393.75  1226583.0  
education-num       10.0      12.00       16.0  
capital-gain         0.0       0.00    99999.0  
capital-loss         0.0       0.00     4356.0  
hours-per-week      40.0      45.00       99.0

In [6]:
income_eval.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

In [7]:
print(income_eval.columns)

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='object')


In [8]:
# Data type conversion from object to string
income_eval[["workclass","education","marital-status","occupation","relationship","race","sex","native-country","income"]] = income_eval[["workclass","education","marital-status",
                                                                                                                                           "occupation","relationship","race","sex","native-country",
                                                                                                                                          "income"]].astype(str)
                                                                                                                                           
#Conversion from object to integer
income_eval[["fnlwgt", "education-num","hours-per-week"]]=income_eval[["fnlwgt", "education-num","hours-per-week"]].astype(int)

In [9]:
income_eval.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

In [10]:
income_eval = income_eval.drop(['fnlwgt','native-country'],axis = 1)

In [11]:
income_eval.head()

age         workclass  education  education-num      marital-status  \
0   39         State-gov  Bachelors             13       Never-married   
1   50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   
2   38           Private    HS-grad              9            Divorced   
3   53           Private       11th              7  Married-civ-spouse   
4   28           Private  Bachelors             13  Married-civ-spouse   

          occupation   relationship   race     sex  capital-gain  \
0       Adm-clerical  Not-in-family  White    Male          2174   
1    Exec-managerial        Husband  White    Male             0   
2  Handlers-cleaners  Not-in-family  White    Male             0   
3  Handlers-cleaners        Husband  Black    Male             0   
4     Prof-specialty           Wife  Black  Female             0   

   capital-loss  hours-per-week income  
0             0              40  <=50K  
1             0              13  <=50K  
2             0              40  <=50K  
3             0              40  <=50K  
4             0              40  <=50K

### Converting the columns to data that can be readable for machine learning
#### **Use make-column-transformer**

In [12]:
from sklearn.preprocessing import OrdinalEncoder

In [13]:
oe = OrdinalEncoder()

In [14]:
income_eval.head(2)

age         workclass  education  education-num      marital-status  \
0   39         State-gov  Bachelors             13       Never-married   
1   50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   

        occupation   relationship   race   sex  capital-gain  capital-loss  \
0     Adm-clerical  Not-in-family  White  Male          2174             0   
1  Exec-managerial        Husband  White  Male             0             0   

   hours-per-week income  
0              40  <=50K  
1              13  <=50K

In [15]:
from sklearn.compose import make_column_transformer

ct = make_column_transformer(
    (oe,['workclass','education','marital-status','occupation','relationship','race','sex']),
    remainder = 'passthrough'
    
)
ct.set_output(transform='pandas')
df_income =ct.fit_transform(income_eval)

In [16]:
df_income.head()

ordinalencoder__workclass  ordinalencoder__education  \
0                        7.0                        9.0   
1                        6.0                        9.0   
2                        4.0                       11.0   
3                        4.0                        1.0   
4                        4.0                        9.0   

   ordinalencoder__marital-status  ordinalencoder__occupation  \
0                             4.0                         1.0   
1                             2.0                         4.0   
2                             0.0                         6.0   
3                             2.0                         6.0   
4                             2.0                        10.0   

   ordinalencoder__relationship  ordinalencoder__race  ordinalencoder__sex  \
0                           1.0                   4.0                  1.0   
1                           0.0                   4.0                  1.0   
2                           1.0                   4.0                  1.0   
3                           0.0                   2.0                  1.0   
4                           5.0                   2.0                  0.0   

   remainder__age  remainder__education-num  remainder__capital-gain  \
0              39                        13                     2174   
1              50                        13                        0   
2              38                         9                        0   
3              53                         7                        0   
4              28                        13                        0   

   remainder__capital-loss  remainder__hours-per-week remainder__income  
0                        0                         40             <=50K  
1                        0                         13             <=50K  
2                        0                         40             <=50K  
3                        0                         40             <=50K  
4                        0                         40             <=50K

### Converting the income column using label encoder

In [17]:
from sklearn.preprocessing import LabelEncoder

In [18]:
le = LabelEncoder()
income_eval['income']=le.fit_transform(income_eval['income'])
income_eval['income'].value_counts()

income
0    7621
1    2379
Name: count, dtype: int64

In [19]:
income_eval.head()

age         workclass  education  education-num      marital-status  \
0   39         State-gov  Bachelors             13       Never-married   
1   50  Self-emp-not-inc  Bachelors             13  Married-civ-spouse   
2   38           Private    HS-grad              9            Divorced   
3   53           Private       11th              7  Married-civ-spouse   
4   28           Private  Bachelors             13  Married-civ-spouse   

          occupation   relationship   race     sex  capital-gain  \
0       Adm-clerical  Not-in-family  White    Male          2174   
1    Exec-managerial        Husband  White    Male             0   
2  Handlers-cleaners  Not-in-family  White    Male             0   
3  Handlers-cleaners        Husband  Black    Male             0   
4     Prof-specialty           Wife  Black  Female             0   

   capital-loss  hours-per-week  income  
0             0              40       0  
1             0              13       0  
2             0              40       0  
3             0              40       0  
4             0              40       0

In [20]:
#Splitting data into X and y
X = df_income.iloc[:,:12]
y = income_eval['income']
print(X.shape)
print(y.shape)


(10000, 12)
(10000,)


In [21]:
#Splitting data into training and splitting datasets
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2 ,random_state=100)

In [22]:
print(X_train.shape)
print(y_train.shape)

(8000, 12)
(8000,)


In [23]:
#Training the model
from sklearn.neighbors import KNeighborsClassifier
model_kn = KNeighborsClassifier(n_neighbors=5)

In [24]:
#Fitting th model
model_kn.fit(X_train,y_train)

KNeighborsClassifier()

In [25]:
#Predicting the model
y_pred= model_kn.predict(X_test)


In [26]:
#Evaluating the accuracy of the model
from sklearn.metrics import accuracy_score,classification_report

In [27]:
print(f"The accuracy of the model is:",accuracy_score(y_pred,y_test))

The accuracy of the model is: 0.832


The K-nearest neighbors preforms better that Support Vector Machine <br>
KNN has an accuracy of 83.2%
SVM has an accuracy of 80.6%